clients/tfchain-client-go/impl.go

```diff
@@ -37,145 +36,437 @@ type Versioned struct {
+	Close() error
+
+	// Deprecated methods
 	Raw() (Conn, Meta, error)
 	Substrate() (*Substrate, error)
 }
 
-type mgrImpl struct {
-	urls []string
+type ManagerConfig struct {
+	// Maximum number of connections in the pool
+	MaxPoolSize int
+	// Minimum number of connections to maintain
+	MinPoolSize int
+	// Maximum time a connection can be idle before being closed (if the pool has more than MinPoolSize)
+	MaxIdleTime time.Duration
+	// Interval between health checks
+	// After thinking about it, we don't need to periodically check the health of the connections
+	// because this creates a lot of overhead
+	// so instead we just check the health when we need to and do the maintanance in demand
+	// HealthCheckInterval time.Duration
+	// Timeout for creating new connections
+	ConnectionTimeout time.Duration
+}
+
+// Default configuration
+var DefaultConfig = ManagerConfig{
+	MaxPoolSize: 5,
+	MinPoolSize: 2,
+	MaxIdleTime: 30 * time.Minute,
+	// HealthCheckInterval: 120 * time.Second,
+	ConnectionTimeout: 10 * time.Second,
+}
+
+type manager struct {
+	urls      []string
+	pool      []*poolConn
+	mu        sync.RWMutex
+	ctx       context.Context
+	cancel    context.CancelFunc
+	wg        sync.WaitGroup
+	config    ManagerConfig
+	checkChan chan struct{}
+}
 
-	r int
-	m sync.Mutex
+func NewManager(urls ...string) Manager {
+	return NewManagerWithConfig(DefaultConfig, urls...)
 }
 
-func NewManager(url ...string) Manager {
-	if len(url) == 0 {
-		panic("at least one url is required")
+func NewManagerWithConfig(config ManagerConfig, urls ...string) Manager {
+	if len(urls) == 0 {
+		panic("at least one URL required")
 	}
 
-	// the shuffle is needed so if one endpoints fails, and the next one
-	// is tried, we will end up moving all connections to the "next" endpoint
-	// which will get overloaded. Instead the shuffle helps to make the "next"
-	// different for reach instace of the pool.
-	rand.Shuffle(len(url), func(i, j int) {
-		url[i], url[j] = url[j], url[i]
-	})
+	// Validate and adjust configuration
+	if config.MaxPoolSize < 1 {
+		config.MaxPoolSize = DefaultConfig.MaxPoolSize
+	}
+	if config.MinPoolSize < 1 || config.MinPoolSize > config.MaxPoolSize {
+		config.MinPoolSize = min(DefaultConfig.MinPoolSize, config.MaxPoolSize)
+	}
+	if config.MaxIdleTime <= 0 {
+		config.MaxIdleTime = DefaultConfig.MaxIdleTime
+	}
+	// if config.HealthCheckInterval <= 0 {
+	// 	config.HealthCheckInterval = DefaultConfig.HealthCheckInterval
+	// }
+	if config.ConnectionTimeout <= 0 {
+		config.ConnectionTimeout = DefaultConfig.ConnectionTimeout
+	}
 
-	return &mgrImpl{
-		urls: url,
-		r:    rand.Intn(len(url)), // start with random url, then roundrobin
+	ctx, cancel := context.WithCancel(context.Background())
+	m := &manager{
+		urls:      shuffle(urls),
+		pool:      make([]*poolConn, 0, config.MaxPoolSize),
+		ctx:       ctx,
+		cancel:    cancel,
+		config:    config,
+		checkChan: make(chan struct{}, 1),
```

Comment:
Why buffered?

Reply:
The buffer size of 1 is sufficient because we only need to know if a check is pending or not.
If a check is already pending, other goroutines are not blocked (Non-blocking sends).
Also, Multiple check requests are coalesced, collapsing multiple requests into a single check operation (reduces contention).
  
```
 select {
    case m.checkChan <- struct{}{}: // Non-blocking send
        log.Debug().Msg("triggered connection check")
    default: // If channel is full, skip
        log.Debug().Msg("connection check already pending")
    }
```

Note that initially, I have the health check execute on demand—immediately when a connection fails—and periodically, however, I later reconsidered the impact of the side effects on the load and the actual benefits.